msg-socket/src/pub/driver.rs

```diff
@@ -0,0 +1,162 @@
+use futures::{Future, SinkExt, StreamExt};
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::{sync::broadcast, task::JoinSet};
+use tokio_util::codec::Framed;
+use tracing::{debug, error};
+
+use super::{
+    session::SubscriberSession, trie::PrefixTrie, PubError, PubMessage, PubOptions, SocketState,
+};
+use crate::{AuthResult, Authenticator};
+use msg_transport::ServerTransport;
+use msg_wire::{auth, pubsub};
+
+pub(crate) struct PubDriver<T: ServerTransport> {
+    /// Session ID counter.
+    pub(super) id_counter: u32,
+    /// The server transport used to accept incoming connections.
+    pub(super) transport: T,
+    /// The publisher options (shared with the socket)
+    pub(super) options: Arc<PubOptions>,
+    /// The publisher socket state, shared with the socket front-end.
+    pub(crate) state: Arc<SocketState>,
+    /// Receiver from the socket front-end.
+    // pub(super) from_socket: mpsc::Receiver<Command>,
+    /// Optional connection authenticator.
+    pub(super) auth: Option<Arc<dyn Authenticator>>,
+    /// A joinset of authentication tasks.
+    pub(super) auth_tasks: JoinSet<Result<AuthResult<T::Io>, PubError>>,
+    /// Stream map of session receivers, keyed by session ID. This is used to listen to
+    /// requests from the session like subscribe, unsubscribe, and to be notified when the
+    /// session is closed (`StreamNotifyClose`).
+    // pub(super) from_sessions: StreamMap<u32, StreamNotifyClose<ReceiverStream<SessionRequest>>>,
+    /// The receiver end of the message broadcast channel. The sender half is stored by [`PubSocket`](super::PubSocket).
+    pub(super) from_socket_bcast: broadcast::Receiver<PubMessage>,
+}
+
+impl<T: ServerTransport> Future for PubDriver<T> {
+    type Output = Result<(), PubError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        loop {
+            if let Poll::Ready(Some(Ok(auth))) = this.auth_tasks.poll_join_next(cx) {
```

Comment:
is there a test/example with pubsub with auth?

Reply:
Not yet, will add!

Reply:
Done in latest pushes

Reply:
💯